src/trie.py

```diff
@@ -46,3 +46,25 @@ def contains(self, token):
             else:
                 return False
         return True
+
+    def traversal(self, start):
+        """
+        Perform a full depth-first traversal of the trie beginning at start.
+        Return a generator containing all tokens in the trie.
+        """
+        def create_token(token, current_dict):
+            for key in current_dict:
+                if key == '$':
+                    yield token
+                else:
+                    token1 = token + key
+                    current_dict1 = current_dict[key]
+                    result = create_token(token1, current_dict1)
+                    for x in result:
+                        yield x
+        token = start
+        if start in self.head.keys():
```

Comment:
This  check is an O(n) operation, because `dict.keys` returns a list in Python 2 and a generator in Python 3. In both scenarios, you have to iterate over the whole thing, hence O(n).

On the other hand, making a key reference on a dictionary is O(1), so it would be more efficient to
```python
try:
    current_dict = self.head[start]
    ...
except KeyError:
    ...
```